Source/Core/Core/HW/DVDThread.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <cinttypes>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "Common/ChunkFile.h"
+#include "Common/CommonTypes.h"
+#include "Common/Event.h"
+#include "Common/Flag.h"
+#include "Common/MsgHandler.h"
+#include "Common/Thread.h"
+#include "Common/Timer.h"
+#include "Common/Logging/Log.h"
+
+#include "Core/Core.h"
+#include "Core/CoreTiming.h"
+#include "Core/HW/DVDInterface.h"
+#include "Core/HW/DVDThread.h"
+#include "Core/HW/Memmap.h"
+#include "Core/HW/SystemTimers.h"
+
+#include "DiscIO/Volume.h"
+
+namespace DVDThread
+{
+
+static void DVDThread();
+static bool ReadFromDisc();
+
+static void FinishRead(u64 userdata, int cyclesLate);
+static int s_finish_read;
+
+static std::thread s_DVD_thread;
+static Common::Event s_DVD_thread_start_working;
+static Common::Event s_DVD_thread_not_working;
+static Common::Flag s_DVD_thread_exiting(false);
+
+static std::vector<u8> s_DVD_buffer;
+static u64 s_time_read_started;
+static bool s_DVD_success;
+
+static u64 s_DVD_offset;
+static u32 s_output_address;
+static u32 s_length;
+static bool s_decrypt;
+
+// Used to notify emulated software after executing command.
+// Pointers don't work with savestates, so CoreTiming events are used instead
+static int s_callback_event_type;
+
+// The following time variables are only used for logging
+static u64 s_realtime_started_us;
+static u64 s_realtime_done_us;
+
+void Start()
+{
+	s_finish_read = CoreTiming::RegisterEvent("FinishReadDVDThread", FinishRead);
+	_assert_(!s_DVD_thread.joinable());
+	s_DVD_thread = std::thread(DVDThread);
+}
+
+void Stop()
+{
+	_assert_(s_DVD_thread.joinable());
+
+	// The DVD thread will return if s_DVD_thread_exiting
+	// is set when it starts working
+	s_DVD_thread_exiting.Set();
+	s_DVD_thread_start_working.Set();
+
+	s_DVD_thread.join();
+
+	s_DVD_thread_exiting.Clear();
+}
+
+void DoState(PointerWrap &p)
+{
+	WaitUntilIdle();
+
+	// TODO: Savestates can be smaller if s_DVD_buffer is not saved
+	p.Do(s_DVD_buffer);
```

Comment:
Maybe just store the command to read the dvd? So you'll just re-read the dvd after loading. It's a bit in contrast to WaitUntilIdle() before, but it may work quite fine.